src/notification/verifyIosNotificationData.js

```diff
@@ -0,0 +1,232 @@
+/* @flow strict-local */
+import type { Notification } from './types';
+import type { JSONable, JSONableDict } from '../utils/jsonable';
+import * as logging from '../utils/logging';
+
+/**
+ * PartialJSONableDict: readonly JSONable dictionary.
+ *
+ * Unlike the definition of `JSONableDict`, this permits the possiblility of
+ * `void` elements. This is a workaround for Flow's by-design soundness hole
+ * involving indexer elements.
+ */
+// eslint-disable-next-line flowtype/type-id-match
+type PartialJSONableDict = { +[string]: void | JSONable };
+
+/** Confirm (or disprove) that a JSONable object is a dictionary. */
+const asDict = (obj: JSONable | void): JSONableDict | void => {
+  if (typeof obj !== 'object') {
+    return undefined;
+  }
+  if (obj === null || obj instanceof Array) {
+    return undefined;
+  }
+  return obj;
+};
+/*
+    The Zulip APNs message format, as far back as 2013-03-23 [0], has always
+    been some subtype of the following:
+
+        type Data = { zulip: { ... } }
+
+    (Comments in `zerver/lib/push_notifications.py` may appear to indicate
+    otherwise, but these refer only to the format of inter-server messages for
+    the Zulip-hosted APNs-bouncer service. Messages sent over APNs itself have
+    always had a `zulip` field.)
+
+    [0] GitHub commit 410ee44eb6e40bac4099d1851d949533026fe4b3, imported from an
+        older commit 6c54fe44a82c5796268e56d3f5577bf4cfc8163a.
+
+
+    This has been expanded incrementally over the years. As of 2020-02, commit
+    2.2-dev-775-g10e7e15088, the current form of APNs messages is as follows:
+
+    ```
+    type StreamData = {
+        // added 1.7.0-1351-g98943a8333, release 1.8.0+
+        recipient_type: 'stream',
+        stream: string,
+        topic: string,
+    };
+
+    type PmData = {
+        // added 1.7.0-1351-g98943a8333, release 1.8.0+
+        recipient_type: 'private',
+
+        // added 1.7.0-2360-g693a9a5e70, release 1.8.0+
+        // present only on huddles (multi-user PMs)
```

Comment:
You say that, but we've got a fair number of occurrences of "huddles" in the client code right now: `UnreadHuddlesState`, `addItemsToHuddleArray`, `getUnreadByHuddles`, etc.

(Altered anyway.)

Reply:
Yeah. Those are all within the `state.unread` reducer code, though, and they're all due to a single such slip-up in the API, in the `unread_msgs` part of the initial data (on `/register`, or in `page_params` for the webapp). That data structure (a) distinguishes group PMs from 1:1 PMs, and (b) uses the term "huddle" for the former, both of which we've otherwise managed to avoid in the API.